crates/ty_python_semantic/src/module_resolver/list.rs

```diff
@@ -0,0 +1,1688 @@
+    /// caller looks for modules in search path priority order.
+    fn add_module(&mut self, path: &ModulePath, module: Module<'db>) {
+        let mut entry = match self.modules.entry(module.name(self.db)) {
+            Entry::Vacant(entry) => {
+                entry.insert(module);
+                return;
+            }
+            Entry::Occupied(entry) => entry,
+        };
+
+        let existing = entry.get();
+        match (existing.search_path(self.db), module.search_path(self.db)) {
+            // When we had a namespace package and now try to
+            // insert a non-namespace package, the latter always
+            // takes precedent, even if it's in a lower priority
+            // search path.
+            (None, Some(_)) => {
+                entry.insert(module);
+            }
+            (Some(search_path_existing), Some(search_path_new)) => {
+                // All of the cases below require that we're
+                // in the same directory.
+                if search_path_existing != search_path_new {
+                    return;
+                }
+                // When we have a `foo/__init__.py` and a `foo.py` in
+                // the same directory, the former takes precedent.
+                // (This case can only occur when both have a search
+                // path.)
+                if existing.kind(self.db) == ModuleKind::Module
+                    && module.kind(self.db) == ModuleKind::Package
+                {
+                    entry.insert(module);
+                    return;
+                }
+                // Or if we have two file modules and the new one
+                // is a stub, then the stub takes priority.
+                if existing.kind(self.db) == ModuleKind::Module
+                    && module.kind(self.db) == ModuleKind::Module
+                    && path.is_stub_file()
+                {
+                    entry.insert(module);
+                    return;
+                }
+                // Or... if we have a stub package, the stub package
+                // always gets priority.
+                if path.is_stub_package() {
+                    entry.insert(module);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    /// Returns true if the given module name cannot be shadowable.
+    fn is_non_shadowable(&self, name: &ModuleName) -> bool {
+        is_non_shadowable(self.python_version().minor, name.as_str())
+    }
+
+    /// Returns the Python version we want to perform module resolution
+    /// with.
+    fn python_version(&self) -> PythonVersion {
+        self.program.python_version(self.db)
+    }
+
+    /// Returns true if and only if the given file extension corresponds
+    /// to a Python source or stub file. This also takes file system
+    /// case sensitivity into account.
+    fn is_python_extension(&self, ext: &str) -> bool {
+        if self.db.system().case_sensitivity().is_case_sensitive() {
```

Comment:
Checking whether the file system is case sensitive can be fairly expensive (as expensive as writing a file to disk!). The best case is that it's already cached. Either way, I'm not sure if this is a worthwhile performance optimization or if we should just always call `eq_ignore_ascii_case`